src/wallet/asyncrpcoperation_saplingmigration.cpp

```diff
@@ -0,0 +1,196 @@
+    {
+        LOCK2(cs_main, pwalletMain->cs_wallet);
+        // Consider, should notes be sorted?
+        pwalletMain->GetFilteredNotes(sproutEntries, saplingEntries, "", 0);
+    }
+    if (sproutEntries.empty()) { // Consider, should the migration remain enabled?
+        pwalletMain->fSaplingMigrationEnabled = false;
+        return true;
+    }
+    CAmount availableFunds = 0;
+    for (const CSproutNotePlaintextEntry& sproutEntry : sproutEntries) {
+        availableFunds = sproutEntry.plaintext.value();
+    }
+    // If the remaining amount to be migrated is less than 0.01 ZEC, end the migration.
+    if (availableFunds < CENT) {
+        pwalletMain->fSaplingMigrationEnabled = false;
+        return true;
+    }
+
+    HDSeed seed;
+    if (!pwalletMain->GetHDSeed(seed)) {
+        throw JSONRPCError(
+            RPC_WALLET_ERROR,
+            "AsyncRPCOperation_AsyncRPCOperation_saplingmigration: HD seed not found");
+    }
+
+    libzcash::SaplingPaymentAddress migrationDestAddress = getMigrationDestAddress(seed);
+
+    auto consensusParams = Params().GetConsensus();
+
+    // Up to the limit of 5, as many transactions are sent as are needed to migrate the remaining funds
+    int numTxCreated = 0;
+    int noteIndex = 0;
+    do {
+        CAmount amountToSend = chooseAmount(availableFunds);
+        auto builder = TransactionBuilder(consensusParams, targetHeight_, pwalletMain, pzcashParams);
+        std::vector<CSproutNotePlaintextEntry> fromNotes;
+        CAmount fromNoteAmount = 0;
+        while (fromNoteAmount < amountToSend) {
+            auto sproutEntry = sproutEntries[noteIndex++];
+            fromNotes.push_back(sproutEntry);
+            fromNoteAmount += sproutEntry.plaintext.value();
+        }
+        availableFunds -= fromNoteAmount;
+        for (const CSproutNotePlaintextEntry& sproutEntry : fromNotes) {
+            libzcash::SproutNote sproutNote = sproutEntry.plaintext.note(sproutEntry.address);
+            libzcash::SproutSpendingKey sproutSk;
+            pwalletMain->GetSproutSpendingKey(sproutEntry.address, sproutSk);
+            std::vector<JSOutPoint> vOutPoints = {sproutEntry.jsop};
+            // Each migration transaction SHOULD specify an anchor at height N-10
+            // for each Sprout JoinSplit description
+            // TODO: the above functionality (in comment) is not implemented in zcashd
+            uint256 inputAnchor;
+            std::vector<boost::optional<SproutWitness>> vInputWitnesses;
+            pwalletMain->GetSproutNoteWitnesses(vOutPoints, vInputWitnesses, inputAnchor);
+            builder.AddSproutInput(sproutSk, sproutNote, vInputWitnesses[0].get());
+        }
+        // The amount chosen *includes* the 0.0001 ZEC fee for this transaction, i.e.
+        // the value of the Sapling output will be 0.0001 ZEC less.
+        builder.SetFee(FEE);
+        builder.AddSaplingOutput(ovkForShieldingFromTaddr(seed), migrationDestAddress, amountToSend - FEE);
```

Comment:
The ZIP should probably specify that `ovk` is generated as if for a transfer from a t-addr.